fgpyo/sam/tests/test_umi_methods.py

```diff
@@ -0,0 +1,127 @@
+from fgpyo.sam import copy_umi_from_read_name
+from fgpyo.sam import extract_umis_from_read_name
+from fgpyo.sam.builder import SamBuilder
+
+
+@pytest.mark.parametrize(
+    "umi,validity",
+    [
+        ("ACGTN", True),
+        ("rACGTN", False),
+        ("+ACGTN", False),
+        ("-ACGTN", False),
+        ("BXCYF", False),
+    ],
+)
+def test_is_valid_umi(umi: str, validity: bool) -> None:
+    """Test that we can detect valid UMIs."""
+
+    assert _is_valid_umi(umi) is validity
+
+
+@pytest.mark.parametrize(
+    "read_name,umi",
+    [
+        ("abc:ACGT", "ACGT"),
+        ("abc:def:ghi:ACGT", "ACGT"),
+        ("abc:def:ghi:rACGT", "ACGT"),
+        ("abc:def:ghi:rACGT+CAGA", "ACGT-CAGA"),
+        ("abc:def:ghi:ACGT+rCAGA", "ACGT-CAGA"),
+        ("abc:def:ghi:rACGT+rCAGA", "ACGT-CAGA"),
+    ],
+)
+def test_extract_umi_from_read_name(read_name: str, umi: str) -> None:
+    """Test that we can extract UMI from a read name."""
+    assert extract_umis_from_read_name(read_name) == umi
+
+
+@pytest.mark.parametrize(
+    "read_name",
+    [
+        "abc:def:ghi:ArCGT",
+        "abc:def:ghi:ACGTr",
+        "abc:def:ghi:ACGT-CAGA",
+        "abc:def:ghi:ACGT_CAGA",
+        "abc:def:ghi:+ACGT",
+        "abc:def:ghi:ACGT+",
+    ],
+)
+def test_extract_umi_from_read_name_raises(read_name: str) -> None:
+    """Test that we raise an error when the read name includes an invalid UMI
+    and strict=True."""
+    with pytest.raises(ValueError):
+        extract_umis_from_read_name(read_name=read_name, strict=True)
+
+
+def test_extract_umi_from_read_name_strict_False() -> None:
+    """Test that we return None when an invalid UMI is encountered
+    and strict=False (but still return a valid UMI)."""
+    assert extract_umis_from_read_name(read_name="abc:def:ghi:ArCGT", strict=False) is None
+    assert extract_umis_from_read_name(read_name="abc:def:ghi:ACGTr", strict=False) is None
+    assert (
+        extract_umis_from_read_name(read_name="abc:def:ghi:rACGT+CAGA", strict=False)
+        == "ACGT-CAGA"
+    )
+
+
+@pytest.mark.parametrize(
+    "read_name, extraction",
+    [
+        ("abc:def:ghi:jfk:lmn:opq:ACGT", None),  # colons == 6
+        ("abc:def:ghi:jfk:lmn:opq:rst:ACGT", "ACGT"),  # colons == 7
+    ],
+)
+def test_strict_extract_umi_from_read_name(read_name: str, extraction: str) -> None:
+    """Test that we raise an error when strict=True and number of colons is not 7 or 8."""
+    assert extract_umis_from_read_name(read_name, strict=True) == extraction
+
+
+@pytest.mark.parametrize(
+    "read_name",
+    [
+        ("abc:def:ghi:jfk"),
+        ("abc:def:ghi:jfk:lmn:opq:rst:uvw:xyz"),
+        ("abc:def:ghi:jfk:lmn:opq:rst:"),  # Invalid UMI
+    ],
+)
+def test_strict_extract_umi_from_read_name_raises(read_name: str) -> None:
+    """Test that we raise an error when strict=True and number of colons is not 7 or 8."""
+    with pytest.raises(ValueError):
+        extract_umis_from_read_name(read_name, strict=True)
```

Comment:
How is this unit test passing as-is?  I think it hits this branch, which doesn't raise but returns:

```python
    elif strict:
        return False
        raise ValueError(f"Invalid UMI {umi} extracted from {rec.query_name}")
```

Reply:
I think this unit-test is overloaded because we have 2 different `ValueError` conditions: one for whether we have 7 or 8 colon-delimited fields, and one for when we encounter an invalid UMI (which is the last condition in the testing input data). The unit test was passing because it was raising a `ValueError`, just not the one that corresponds to the code snippet you included (it instead raised an earlier `ValueError` corresponding to `f"Trying to extract UMIs from read with {colons + 1} parts, (7 or 8 expected): {read_name}"`).

I separated out the two input data conditions and also added a `match` to the raise condition so we can track exactly which `ValueError` is getting triggered.